### ISSUE TITLE
Bind axis + movement to virtual button A and - movement to virtual button B.

### ISSUE THREAD

**Issue by bluddy** [WhiteMagic/JoystickGremlin#256]
I have 2 axis dials on my HOTAS, and for most games, they're not necessary. I could turn them into virtual buttons with ranges, but I'd rather have them map to two buttons each: one button when dialed up, and one when dialed down. I prefer buttons because that gives me flexibility to map them easily in-game. I don't want the button to be pressed constantly, as that is a very uncommon requirement in games.

The only way I can think of doing that right now is creating many thin ranges on the axes and remapping them to a virtual button. But that's a lot of work (and may cause slowdown?) and this is a very simple use case. Also, if I end up moving onto a range, I'll get stuck with the button in the on position, which is rarely supported in games, as mentioned above. As an example of how common this request it, it's actually what the dosbox keymapper does by default: it can map an axis' positive movement to one key/button, and negative movement to another key/button. I'd love to see this supported on Gremlin.

**Comment by WhiteMagic** [comment]
There is no functionality in Gremlin to do this directly via the UI, as the behavior you're after requires some memory to figure out if a value increased or decreased. However, writing a plugin that implements this functionality should be rather easy. The one thing I'm not sure from the top of my head is if I have a variable that accepts keyboard keys or combinations thereof as input. If that doesn't exist it's a good idea to add onto the todo list and the workaround would be to have vJoy buttons be used instead.

**Comment by bluddy** [comment]
Yeah I'm happy to use vJoy buttons -- in fact I prefer it, as they can then be mapped by any game to the desired functionality.

Is there a guide to writing plugins, or specifically to the API?

Even though this requires one float of memory per axis, I think this is probably a very popular request and it's worth considering its integration into the UI. For example, older games don't have support for a throttle axis, but by mapping `+` and `-` to throttle increase and decrease, you can get decent functionality (if supported by the game, you could then add extreme points with absolute values such as 'full throttle' to make sure you don't get stuck in mid-range).

**Comment by WhiteMagic** [comment]
There isn't really a guide on those plugins yet, though if you want to have an idea of what they look like you can have a look at https://whitemagic.github.io/JoystickGremlin/user_plugins_code/ Though I probably should find the time over the weekend to write a plugin that implements that functionality.

Adding it as an action of its own is a possibility, though I probably wouldn't integrate it into an existing one to prevent cluttering the UI ... though since all that is being redone who knows.

Ages ago I wrote a plugin that did map a throttle for a game where you could only control it via +/- keys. It did some hackery to ensure that the throttle setting matched the ingame throttle by resetting it and applying the correct amount of key presses. That code https://gist.github.com/WhiteMagic/96f2fa7ca32b33e9cf95e796fa1dc426 isn't configurable via the plugin system at the moment, however.

**Comment by bluddy** [comment]
Just to give a practical example, the Star Wars Tie Fighter and X-Wing games have bindings for 0%, 33%, 66% and 100% throttle. They also have +3% and -3% throttle. (The DOS version has no throttle axis support, and the Windows version was a lazy port and only bound the axis to those 4 absolute points.) My plan was to eventually map those 0, 33, 66, and 100 points as actual small ranges on the axis, so that when hitting those points, you automatically get set to the appropriate absolute setting, but also to create long ranges between those points where every move up or down will nudge + or - 3%, respectively. The fact that most of the movement is relative doesn't matter since hitting the absolute points will reset to the appropriate value.

**Comment by WhiteMagic** [comment]
Here is a simple plugin that makes an axis emit a button tap whenever a specified amount of travel has occurred.

https://gist.github.com/WhiteMagic/ff636574b45a12480fd0bd6f1a92740e

I think if you combine this with the virtual buttons on an axis you should get the behavior you described above.
